engine/rootmap.3
.\"	$Csoft: rootmap.3,v 1.8 2004/05/10 05:14:52 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 23, 2002
.Dt ROOTMAP 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm rootmap
.Nd agar tile-based map renderer
.Sh SYNOPSIS
.Bd -literal
#include <engine/view.h>
#include <engine/map.h>
#include <engine/rootmap.h>
.Ed
.Sh DESCRIPTION
The agar
.Nm
interface manages a single tile-based
.Xr map 3
display.
.Nm
functions are only usable after a call to
.Xr view_init 3
with the
.Dv GFX_ENGINE_TILEBASED
argument.
.Sh MAP DISPLAY
.nr nS 1
.Ft "void"
.Fn rootmap_redraw "void"
.Pp
.Ft "void"
.Fn rootmap_animate "void"
.Pp
.Ft "void"
.Fn rootmap_draw "void"
.Pp
.Ft "void"
.Fn rootmap_focus "struct map *map"
.Pp
.Ft "void"
.Fn rootmap_center "struct map *map" "int x" "int y"
.Pp
.Ft "void"
.Fn rootmap_scroll "struct map *map" "int direction"
.nr nS 0
.Pp
The
.Fn rootmap_redraw
function is called to set the redraw flag on the root map.
.Pp
The
.Fn rootmap_animate
function updates the animations displayed by the root map.
.Fn rootmap_draw
renders static nodes.
.Fn rootmap_draw
is usually called from the event loop
when the
.Dv redraw
flag of the root map is non-zero.
.Pp
The
.Fn rootmap_focus
function displays the given
.Fa map .
The
.Fn rootmap_center
function centers the view on coordinates
.Fa x
and
.Fa y .
The
.Fn rootmap_scroll
function moves the view in the given
.Fa direction .
.Sh SEE ALSO
.Xr agar 3 ,
.Xr map 3 ,
.Xr view 3
.Sh HISTORY
The
.Nm
interface first appeared in Agar 1.0


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

engine/rootmap.c
/*	$Csoft: rootmap.c,v 1.25 2003/03/12 07:59:00 vedge Exp $	*/

/*
 * Copyright (c) 2002, 2003 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "engine.h"

#include "map.h"
#include "rootmap.h"
#include "physics.h"
#include "view.h"

#ifdef DEBUG
#define DEBUG_FOCUS	0x01
#define DEBUG_DIAG	0x02

int	rootmap_debug = DEBUG_DIAG|DEBUG_FOCUS;
#define	engine_debug rootmap_debug
#endif

void
rootmap_init(struct viewmap *rm, int mw, int mh)
{
	rm->w = mw - 1;
	rm->h = mh - 1;
	rm->map = NULL;
	rm->x = 0;
	rm->y = 0;
	rm->sx = 0;
	rm->sy = 0;
	
	/* Calculate the default tile coordinates. */
	rm->maprects = rootmap_alloc_maprects(mw, mh);
}

/*
 * Render all animations in the map view.
 * The view and the displayed map must be locked.
 */
void
rootmap_animate(void)
{
	struct map *m = view->rootmap->map;
	struct viewmap *rm = view->rootmap;
	struct node *node;
	struct noderef *nref;
	int mx, my, rx, ry;
	int layer = 0;

#ifdef DEBUG
	/* XXX */
	if (rm->x < 0 || rm->y < 0 ||
	    rm->y > m->maph - rm->h || rm->x > m->mapw - rm->w) {
		fatal("bad offs");
	}
#endif

draw_layer:
	if (!m->layers[layer].visible) {
		goto next_layer;
	}
	for (my = rm->y, ry = rm->sy - m->tileh;	/* Downward */
	     (my - rm->y) < m->maph && my < m->maph;
	     my++, ry += m->tileh) {

		for (mx = rm->x, rx = rm->sx - m->tilew; /* Forward */
		     (mx - rm->x) < m->mapw && mx < m->mapw;
		     mx++, rx += m->tilew) {
			node = &m->map[my][mx];
			MAP_CHECK_NODE(node);

			TAILQ_FOREACH(nref, &node->nrefs, nrefs) {
				MAP_CHECK_NODEREF(nref);
				if (nref->type == NODEREF_ANIM &&
				    nref->layer == layer) {
					noderef_draw(m, nref, rx, ry);
				}
			}
		}
	}
next_layer:
	if (++layer < m->nlayers) {
		goto draw_layer;			/* Draw next layer */
	}
}

/*
 * Draw static tiles.
 * The view and the displayed map must be locked.
 */
void
rootmap_draw(void)
{
	struct map *m = view->rootmap->map;
	struct viewmap *rm = view->rootmap;
	struct node *node;
	struct noderef *nref;
	int mx, my, rx, ry;
	int layer = 0;

#ifdef DEBUG
	/* XXX */
	if (rm->y > m->maph - rm->h || rm->x > m->mapw - rm->w) {
		debug(DEBUG_DIAG, "offset exceeds map boundaries\n");
		return;
	}
#endif

draw_layer:
	if (!m->layers[layer].visible) {
		goto next_layer;
	}
	for (my = rm->y, ry = rm->sy - m->tileh;		/* Downward */
	     (my - rm->y) < m->maph && my < m->maph;
	     my++, ry += m->tileh) {

		for (mx = rm->x, rx = rm->sx - m->tilew;	/* Forward */
		     (mx - rm->x) < m->mapw && mx < m->mapw;
		     mx++, rx += m->tilew) {
			node = &m->map[my][mx];
			MAP_CHECK_NODE(node);

			TAILQ_FOREACH(nref, &node->nrefs, nrefs) {
				MAP_CHECK_NODEREF(nref);
				if (nref->type == NODEREF_SPRITE &&
				    nref->layer == layer) {
					noderef_draw(m, nref, rx, ry);
				}
			}
		}
	}
next_layer:
	if (++layer < m->nlayers) {
		goto draw_layer;			/* Draw next layer */
	}
}

void
rootmap_focus(struct map *m)
{
	pthread_mutex_lock(&view->lock);
	debug(DEBUG_FOCUS, "%s -> %s\n",
	    (view->rootmap == NULL || view->rootmap->map == NULL) ? "NULL" :
	    OBJECT(view->rootmap->map)->name, OBJECT(m)->name);
	view->rootmap->map = m;
	pthread_mutex_unlock(&view->lock);
}

void
rootmap_center(struct map *m, int mapx, int mapy)
{
	struct viewmap *rm = view->rootmap;
	int nx, ny;

	debug(DEBUG_FOCUS, "centering on map `%s'\n", OBJECT(m)->name);

	pthread_mutex_lock(&m->lock);
	
	nx = mapx - rm->w/2;
	ny = mapy - rm->h/2;

	if (nx < 0)
		nx = 0;
	if (ny < 0)
		ny = 0;
	if (nx >= m->mapw - rm->w)
		nx = m->mapw - rm->w;
	if (ny >= m->maph - rm->h)
		ny = m->maph - rm->h;

	rm->x = nx;
	rm->y = ny;
	m->redraw++;
	
	pthread_mutex_unlock(&m->lock);
}

void
rootmap_scroll(struct map *m, int dir, int inc)
{
	struct viewmap *rm;
	
	pthread_mutex_lock(&view->lock);

	rm = view->rootmap;

	switch (dir) {
	case DIR_UP:
		if ((rm->sy -= inc) <= -TILEH) {
			if (--rm->y < 0) {
				rm->y = 0;
			}
			rm->sy = 0;
		}
		break;
	case DIR_DOWN:
		if ((rm->sy += inc) >= TILEH) {
			if (++rm->y > (m->maph - rm->h)) {
				rm->y = m->maph - rm->h;
			}
			rm->sy = 0;
		}
		break;
	case DIR_LEFT:
		if ((rm->sx -= inc) <= -TILEW) {
			if (--rm->x < 0) {
				rm->x = 0;
			}
			rm->sx = 0;
		}
		break;
	case DIR_RIGHT:
		if ((rm->sx += inc) >= TILEH) {
			if (++rm->x > (m->mapw - rm->w)) {
				rm->x = m->mapw - rm->w;
			}
			rm->sx = 0;
		}
		break;
	}
	m->redraw++;
	
	pthread_mutex_unlock(&view->lock);
}

SDL_Rect **
rootmap_alloc_maprects(int w, int h)
{
	SDL_Rect **rects;
	int x, y;

	/* Calculate the default coordinates of visible rectangles. */
	rects = emalloc(h * sizeof(SDL_Rect *));
	for (y = 0; y < h; y++) {
		rects[y] = emalloc(w * sizeof(SDL_Rect));
		for (x = 0; x < w; x++) {
			rects[y][x].x = x * TILEW;
			rects[y][x].y = y * TILEH;
			rects[y][x].w = TILEW;
			rects[y][x].h = TILEH;
		}
	}
	return (rects);
}

void
rootmap_free_maprects(struct viewport *v)
{
	int y;

	pthread_mutex_lock(&v->lock);
	for (y = 0; y < v->rootmap->h; y++) {
		free(v->rootmap->maprects[y]);
	}
	free(v->rootmap->maprects);
	pthread_mutex_unlock(&v->lock);
}

